File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/SquelchCommand.cs
﻿using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Net;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class SquelchCommand : ChatCommand
    {
        public SquelchCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            string r; // reply
            string t; // target

            t = Arguments.Count == 0 ? "" : Arguments[0];

            if (!Battlenet.Common.GetClientByOnlineName(t, out var target) || target == null)
            {
                r = Resources.UserNotLoggedOn;
                foreach (var line in r.Split(Environment.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            var ipAddress = IPAddress.Parse(target.Client.RemoteEndPoint.ToString().Split(':')[0]);

            lock (context.GameState.SquelchedIPs)
            {
                if (!context.GameState.SquelchedIPs.Contains(ipAddress))
                {
                    context.GameState.SquelchedIPs.Add(ipAddress);
                }
            }

            lock (context.GameState.ActiveChannel)
            {
                if (context.GameState.ActiveChannel != null) context.GameState.ActiveChannel.SquelchUpdate(context.GameState);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/SquelchCommand.vb
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.Net

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class SquelchCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim r As String
            Dim t As String
            t = If(Arguments.Count = 0, "", Arguments(0))
            Dim target As GameState = Nothing

            If Not Battlenet.Common.GetClientByOnlineName(t, target) OrElse target Is Nothing Then
                r = Resources.UserNotLoggedOn

                For Each line In r.Split(Environment.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            Dim rereIpAddress = IPAddress.Parse(target.Client.RemoteEndPoint.ToString().Split(":"c)(0))

            SyncLock varContext.GameState.SquelchedIPs

                If Not varContext.GameState.SquelchedIPs.Contains(rereIpAddress) Then
                    varContext.GameState.SquelchedIPs.Add(rereIpAddress)
                End If
            End SyncLock

            SyncLock varContext.GameState.ActiveChannel
                If varContext.GameState.ActiveChannel IsNot Nothing Then varContext.GameState.ActiveChannel.SquelchUpdate(varContext.GameState)
            End SyncLock
        End Sub
    End Class
End Namespace
